dali/pipeline/operators/arithmetic/arithmetic_meta.h

```diff
@@ -0,0 +1,266 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_OPERATORS_ARITHMETIC_ARITHMETIC_META_H_
+#define DALI_PIPELINE_OPERATORS_ARITHMETIC_ARITHMETIC_META_H_
+
+#include <cstdint>
+#include <type_traits>
+
+#include "dali/core/cuda_utils.h"
+#include "dali/core/static_switch.h"
+#include "dali/pipeline/data/types.h"
+
+namespace dali {
+
+enum class ArithmeticOp {
+  id,
+  neg,
+  add,
+  sub,
+  mul,
+  div,
+  mod,
+};
+
+constexpr int GetOpArity(ArithmeticOp op) {
+  auto op_v = static_cast<int>(op);
+  if (static_cast<int>(ArithmeticOp::id) <= op_v && op_v <= static_cast<int>(ArithmeticOp::neg)) {
+    return 1;
+  }
+  if (static_cast<int>(ArithmeticOp::add) <= op_v && op_v <= static_cast<int>(ArithmeticOp::div)) {
+    return 2;
+  }
+  return -1;
+}
+
+/**
+ * @brief Type promotion rules
+ *
+ * @tparam L
+ * @tparam R
+ *
+ * The ruls are:
+ * T      + T      = T
+ * floatX + T      = floatX           (where T is not a float)
+ * floatX + floatY = float(max(X, Y))
+ * intX   + intY   = int(max(X, Y))
+ * uintX  + uintY  = uint(max(X, Y))
+ * intX   + uintY  = int2Y            (if X <= Y)
+ * intX   + uintY  = intX             (if X > Y)
+ *
+ * Operating on integer types never leaves the integer types, and caps at [u]int64
+ *
+ * @return type - a result type of binary arithmetic operation between L and R
+ */
+template <typename L, typename R>
+struct binary_op_promotion;
```

Comment:
I suggest `static_assert` for `is_fundamental` for `L` and `R`. As a sanity check

Reply:
Won't be that misleading? Better fail for anything that is not registered.